### PR TITLE
Upgrade externally managed content

### PR DESCRIPTION
**Pull request by matentzn**
Release Update
- [x] Get Synonym Sync branch and Week 4 steps into `main`
- [ ] Update this branch to point to `main` and confirm build still runs

---
## Overview

This PR:
- Adds all missing pieces of externally managed content to Mondo Ingest (EFO, ClinGen, MedGen)
- And make goals to update them

### Documentation
Was the documentation added/updated under `docs/`?
- [x] Yes
- [ ] No, updates to the docs were not necessary after careful consideration

### QC
Was the full pipeline run before submitting this PR using `sh run.sh make build-mondo-ingest` on this branch (after 
`docker pull obolibrary/odkfull:dev`), and no errors occurred?
- [ ] Yes
- [ ] No, there are no functional (code-related) changes to the pipeline in the PR, so no re-run is necessary
   
### New Packages
Were any new *Python packages* added?
- [ ] Yes, [requirements.txt.full](https://github.com/INCATools/ontology-development-kit/blob/master/requirements.txt.full) was updated and an [ODK issue](https://github.com/INCATools/ontology-development-kit/issues/new) submitted
- [X] No

Were any other *non-Python packages* added?
- [ ] Yes, [Dockerfile](https://github.com/INCATools/ontology-development-kit/blob/master/Dockerfile) updated and an [ODK issue](https://github.com/INCATools/ontology-development-kit/issues/new) submitted
- [X] No

### PR Review and Conversations Resolved
Has the PR been sufficiently reviewed by at least 1 team member of the Mondo Technical team and all threads resolved?
- [ ] Yes

### CODE REVIEW

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:675]: This comment is above `efo-proxy-merges.robot.owl`, but it says TSV; Should it be on the line above `mondo-efo.robot.tsv` or `mondo-otar-subset.robot.tsv`?

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:678]: Usually `robot.owl`s are created from `robot.tsv`s. It makes me wonder if the manually curated file was a`efo-proxy-merges.robot.tsv`? I do see that `efo-proxy-merges.robot.owl` says "Generated by the OWL API".

I suppose we don't expect the `robot.tsv` to ever change, so you are excluding it and simply including the `robot.owl` statically?
If so, then why is there a goal for it? Is it simply so that we see this warning appear in the log, and no other reason?

File: src/ontology/external/efo-proxy-merges.robot.owl
Comment (joeflack4) [inline src/ontology/external/efo-proxy-merges.robot.owl:1]: Curious to learn more about this file; I can learn more about it in time.

File: src/ontology/mondo-ingest.Makefile
Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:675]: No I meant: the TSV file _corresponding_ to the `efo-proxy-merges.robot.owl` file. There must have been one in distant past (in externally managed content, all primary data is in TSV format, and OWL is derived automatically with robot. That file is checked in in OWL).

File: src/ontology/mondo-ingest.Makefile
Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:678]: > It makes me wonder if the manually curated file was aefo-proxy-merges.robot.tsv? I do see that efo-proxy-merges.robot.owl

Yes, see above. Probably.

> I suppose we don't expect the robot.tsv to ever change, so you are excluding it and simply including the robot.owl statically?

The TSV does not exist at all, see comment above.

> If so, then why is there a goal for it? Is it simply so that we see this warning appear in the log, and no other reason?

Yes, just to have a record. Its easier to debug it like this, and for a developer to see when the pipeline is updated.

File: src/ontology/external/efo-proxy-merges.robot.owl
Comment (matentzn) [reply src/ontology/external/efo-proxy-merges.robot.owl:1]: What this file does is to document cases where 2 efo classes are linked to a single Mondo ID (basically adding a "preferredExternal" annotation. We should make sure to document this file before merging this PR, so dont resolve your comment here for now.

**Comment by matentzn** [comment]
Next steps:

1. @twhetzel to update develop from main
2. @twhetzel to update `externalclingenmedgenefo` from develop
3. @twhetzel to update `mondo-ingest/src/ontology/config/robot_report_external_content.txt` with all important checks she can think of
4. @twhetzel to run full build of pipeline and _commit changes to all files in `src/ontology/external` to this branch (only those files)
5. @twhetzel to approve PR and merge
6. @matentzn to implement update EMC pipeline in Mondo repo

@twhetzel agreed?

File: src/ontology/lexmatch/README.md
Comment (joeflack4) [inline src/ontology/lexmatch/README.md:1]: `nit`: Would be nice to separate the build PR from this one so that code changes are easier to review.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile:600]: I just realized that in the `git` diff it looks like you had more than 2 tabs; just looked like more. In any case the 2 examples above are just 1 tab.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:617]: Design `nit`: In other pipelines, we've frowned on the method of a CLI where we just pass the ID of a source, and have preferred to list all of the source-related inputs and outputs as individual params.

I don't lean as strongly towards this preference though, and I understand why @matentzn went with this approach instead as it is much quicker to implement / maintain.

File: src/scripts/post_process_externally_managed_content.py
Comment (joeflack4) [inline src/scripts/post_process_externally_managed_content.py:106]: These look like they should be 'info'. Is `logger.error()` temporary?

File: src/scripts/post_process_externally_managed_content.py
Comment (joeflack4) [inline src/scripts/post_process_externally_managed_content.py:7]: `nit` / question on Python code style: Just highlighting this line, as typically between functions we have 2 line breaks.

I haven't been using `black` for `mondo-ingest`, just following PyCharm's defaults. I honestly don't mind right now if you / we don't follow any single style standard for `mondo-ingest` until we have more time. But just curious if you are planning to use `black` or anything else for your work in `mondo-ingest`.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:613]: Got: `make: *** No rule to make target 'mondo.owl', needed by 'tmp/mondo-incl-external.owl'.  Stop.`
Due to pathing issue ([comment](https://github.com/monarch-initiative/mondo-ingest/pull/628/#discussion_r1722389110))

Trying to spend a limited amount of time reviewing this so I'll skip for now.

File: src/scripts/post_process_externally_managed_content.py
Comment (joeflack4) [inline src/scripts/post_process_externally_managed_content.py:8]:
```suggestion
def is_qc_failure_related_to_external_source(qc_failure, erroneous_row, external) -> Union[bool, str]:
```

### Function name & typing: `is_qc_failure_related_to_external_source()`
I see this is related to the cool system you guys have put together for reporting QC failures in `mondo-ingest` as mentioned on [slack](https://monarchinitiative.slack.com/archives/G01JACCEDAB/p1724105022538769).

Just noting that this was confusing to me. Because of `is_` at the beginning of the function name, I at first expected this to only return a boolean.

I'm suggesting `Union[bool, str]` return type, but you can also make it just `str` by returning `''` in the case of no matches instead of `False`.

Really though this function should be renamed, but I'm not yet sure what to.

File: src/scripts/post_process_externally_managed_content.py
Comment (joeflack4) [inline src/scripts/post_process_externally_managed_content.py:63]: ## `drop_rows_indexes` empty?
I don't see this PR in draft mode, so I would expect this Python script to be complete. However, I'm guessing it is not because `drop_rows_indexes` is empty and never gets updated.

Comment (joeflack4) [inline src/scripts/post_process_externally_managed_content.py:63]: ## Purpose & naming
"Processed" and "post-processing' is a very general words, and since this function and script don't have docstrings, then without looking closely at the code I do not know what kind of processing happening.

Correct me if I'm wrong, but it looks like the entire purpose is to filter out erroneous cells and/or rows in ROBOT templates which would otherwise lead to QC issues in `mondo`, as mentioned on [slack](https://monarchinitiative.slack.com/archives/G01JACCEDAB/p1724105022538769).

File: src/scripts/post_process_externally_managed_content.py
Comment (joeflack4) [inline src/scripts/post_process_externally_managed_content.py:1]: ## General review
This is a lot of work / new content; this is pretty awesome.

My review is a little limited in that I don't have much previous knowledge about the details of EFO. I realize I'm not the primary reviewer, but for me if I have an issue to reference or some more commentary in the OP, or some other resource linked, I could have some better context.
Actually as I'm writing this, I just searched Google Drive and I do see [a document about EMC](https://docs.google.com/document/d/1SLCIIxbW-B-cXfg15A6Xpq9oQ9t0yaHJ_yVkaQia__c/edit) which I'll a skim when I have some time or otherwise go over with Trish.

In any case, I see that each file has custom things we're importing, such as xrefs, synonyms, or gene information.

In many / most cases, it looks like given that these are externally managed, in many cases we don't even have to construct ROBOT templates as that has already been done.

**Approval**
I think my approval is contingent not on design or implementation, etc, but just on making sure that there are no gross errors. As it stands right now, there are just a few things that look like the pipeline is not yet running correctly: [1](https://github.com/monarch-initiative/mondo-ingest/pull/628/#discussion_r1722391864), [2](https://github.com/monarch-initiative/mondo-ingest/pull/628/#discussion_r1722456009), [3](https://github.com/monarch-initiative/mondo-ingest/pull/628/#discussion_r1722389110)

File: src/ontology/config/robot_report_external_content.txt
Comment (matentzn) [reply src/ontology/config/robot_report_external_content.txt:5]: No, this file has to be manually curated by Trish - its on the pending items on this PR.

File: src/ontology/mondo-ingest.Makefile
Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:603]: Done!

File: src/ontology/lexmatch/README.md
Comment (matentzn) [reply src/ontology/lexmatch/README.md:1]: This was not the case, it just seemed this way - the PR was made made of a branch off main which was ahead of develop by one data release. This has now been fixed.

File: .gitignore
Comment (matentzn) [reply .gitignore:114]: It was not necessary, a temporary issue. Thx, removed now.

File: src/scripts/post_process_externally_managed_content.py
Comment (matentzn) [reply src/scripts/post_process_externally_managed_content.py:106]: Thanks for pointing that out, fixed.

File: src/ontology/mondo-ingest.Makefile
Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:616]: I think this is fixed now, a remnant from testing. Thanks for pointing that out!

File: src/ontology/mondo-ingest.Makefile
Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:620]: You are right. Thanks for pointing that out. I changed the script to take only one TSV file as input, and write 1 tsv file as output.

Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:620]: (I like being very explicit on IO parameters in pipelines)

File: src/scripts/post_process_externally_managed_content.py
Comment (matentzn) [reply src/scripts/post_process_externally_managed_content.py:7]: I would like to but not without a automated linting system. I made this one change now, but anything else - only after mondolib is in existence.

File: src/scripts/post_process_externally_managed_content.py
Comment (matentzn) [reply src/scripts/post_process_externally_managed_content.py:8]: I agree with the function name - it used to be a boolen method, but it evolved. Now I changed it to `get_column_of_external_source_related_to_qc_failure`. I didnt add your annotations because if we go this way we need to add them in 100 places - no time!

File: src/scripts/post_process_externally_managed_content.py
Comment (matentzn) [reply src/scripts/post_process_externally_managed_content.py:63]: Yes, I changed the name now to `remove_erroneous_values_from_externally_managed_content`.. Thank you!

Comment (matentzn) [reply src/scripts/post_process_externally_managed_content.py:63]: OOps yes, nothing wild but this was part of a previous version of the script. Dropped this now!

File: src/scripts/post_process_externally_managed_content.py
Comment (matentzn) [reply src/scripts/post_process_externally_managed_content.py:1]: Thank you @joeflack4 I fixed all the issues you uncovered! Much appreciated!

File: src/ontology/medgen-xrefs.robot.template.tsv
Comment (matentzn) [inline src/ontology/medgen-xrefs.robot.template.tsv:1]: WRONG DIR

File: src/ontology/mondo-ingest.Makefile
Comment (matentzn) [inline src/ontology/mondo-ingest.Makefile:614]: DONT FORGET

File: src/ontology/config/robot_report_external_content.txt
Comment (twhetzel) [reply src/ontology/config/robot_report_external_content.txt:5]: Yes, I have an update I'll add in the next ~1 hr.

Comment (twhetzel) [reply src/ontology/config/robot_report_external_content.txt:5]: Check for proxy merges added now.

Comment (joeflack4) [reply src/ontology/config/robot_report_external_content.txt:5]: Oh, ok. This isn't an accidental instance of errors being piped into this file. This is actually a config file representing errors we are trying to account for / filter out of our templates, yes? I'm unfamiliar with these `--profile`s.

Should the file not be a `.tsv` instead of `.txt`?
It looks like it has 2 columns. I'm supposing these profile configs can't have column headers on them? If they did it'd be easier for a lay person to understand the file.

File: src/scripts/post_process_externally_managed_content.py
Comment (joeflack4) [reply src/scripts/post_process_externally_managed_content.py:1]: Wow, I looked through resolved threads; you addressed everything, even the optional stuff.

I see you changed some funcs to private, that's good.

I see script name is still general "post processing" but I think that's good because the func script does slightly more than just filter errors, and it could also be further extended in the future.

Regarding typedefs, I'm fine w/ you not using them, but IMO we don't have to go straight to 100% coverage from 0%. Almost all of my funcs/methods have em for params/returns. Many parts of the codebase have them and many don't.

I'm approving now; will keep a look out for the build.

**Review by joeflack4 — APPROVED** [review]

**Comment by matentzn** [comment]
@twhetzel I am running a testbuild now, and I have also written some docs that explain the process.

**Comment by matentzn** [comment]
@twhetzel back to you! Review needed, of this PR and #644

**Comment by twhetzel** [comment]
The [data build](https://github.com/monarch-initiative/mondo-ingest/pull/644) looks good.

**Comment by twhetzel** [comment]
@matentzn as we discussed earlier, the qc check for `tmp/mondo/src/sparql/qc/mondo/qc-animal-disease-rare.sparql` is not working correctly. Otherwise, all other checks passed (https://github.com/monarch-initiative/mondo/pull/8095).

**Review by twhetzel — CHANGES_REQUESTED** [review]
QC check for animal disease in rare subset is not working yet.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:403]: I think this comment is not supposed to be on `$(TMPDIR)/mondo.sssom.tsv`, but perhaps `$(TMPDIR)/mondo-with-simulated-emc-run.owl`.

Also IDK if `make` `--dry-run` would be helpful at all here. Just throwing that out there since I think we've underutilized it.

Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile]: @joeflack4 thanks for the catch, moved the comment to the right place and improved it a bit.

Not sure about your --dry-run comment in this context! Can you expand what you are suggesting?

**Review by matentzn — COMMENTED** [review]
I have one change request (related to EFO proxy merges, see comments) which I would like your opinion on @twhetzel, else I think this is good to go. You can run the pipeline with:

```
sh run.sh make update-externally-managed-content -B
```

However, it does take a long time, so strap in. The biggest downside of this PR is the fact that I chose to properly simulate the EMC pipeline, which adds a big overhead on the runtime.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile]: Yeah I'm not sure either without understanding what your simulation command is exactly doing.

But basically what `--dry-run` does is let you see all of the goals (the dependency chain) that will run as a result of you running `make GOAL` without actually executing them. Maybe that's not helpful here, I'm not sure.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:390]: 👍

File: src/ontology/mondo-ingest.Makefile
Comment (twhetzel) [reply src/ontology/mondo-ingest.Makefile:713]: @matentzn I agree that we should not fix these on our side, but work with the provider of the source to fix these or a a minimum have the provider determine which should be tagged as preferred.

File: src/ontology/mondo-ingest.Makefile
Comment (twhetzel) [inline src/ontology/mondo-ingest.Makefile:624]: where is this goal in the Makefile?

**Review by twhetzel — APPROVED** [review]
I will Approve for now, but I do think that proxy merges from EFO should not be fixed automatically related to [this comment](https://github.com/monarch-initiative/mondo-ingest/pull/628/files#r1731876897). I don't know if there is still time to change this.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile:624]: Looks like it has already made its way into [`mondo` `main` makefile](https://github.com/monarch-initiative/mondo/blob/master/src/ontology/mondo.Makefile#L316).

Comment (matentzn) [reply src/ontology/mondo-ingest.Makefile:624]: Yes, all that you see here is in the context of the `mondo`, not `mondo-ingest` pipeline!

**Comment by twhetzel** [comment]
The full build does not work for this branch. I am looking into this.

Error:
```
if [ false = true ]; then wget "https://raw.githubusercontent.com/monarch-initiative/mondo-ingest/externalclingenmedgenefo/src/ontology/external/processed-mondo-efo.robot.owl" -O tmp/external/processed-mondo-efo.robot.owl; fi
mkdir -p tmp/external
if [ false = true ]; then wget "https://raw.githubusercontent.com/monarch-initiative/mondo-ingest/externalclingenmedgenefo/src/ontology/external/processed-efo-proxy-merges.robot.owl" -O tmp/external/processed-efo-proxy-merges.robot.owl; fi
make[4]: Leaving directory '/work/src/ontology/tmp/mondo/src/ontology'
grep -vE '^(xref: EFO:|subset: otar)' mondo-edit.obo > tmp/mondo-edit.tmp || true
mv tmp/mondo-edit.tmp mondo-edit.obo
robot --catalog catalog-v001.xml merge -i mondo-edit.obo -i tmp/external/processed-mondo-otar-subset.robot.owl -i tmp/external/processed-mondo-efo.robot.owl -i tmp/external/processed-efo-proxy-merges.robot.owl --collapse-import-closure false \
	query --use-graphs false --update ../sparql/update/update-equivalent-obsolete.ru \
	convert -f obo --check false -o mondo-edit.obo.obo
tmp/external/processed-efo-proxy-merges.robot.owl (No such file or directory)
Use the -vvv option to show the stack trace.
Use the --help option to see usage information.
make[3]: *** [mondo.Makefile:460: update-efo-subset] Error 1
make[3]: Leaving directory '/work/src/ontology/tmp/mondo/src/ontology'
make[2]: *** [mondo.Makefile:319: update-external-content-incl-rare] Error 2
make[2]: Leaving directory '/work/src/ontology/tmp/mondo/src/ontology'
make[1]: *** [mondo-ingest.Makefile:617: tmp/mondo-with-simulated-emc-run.owl] Error 2
rm imports/ro_terms_combined.txt
make[1]: Leaving directory '/work/src/ontology'
make: *** [mondo-ingest.Makefile:346: build-mondo-ingest] Error 2
Command exited with non-zero status 2
### DEBUG STATS ###
Elapsed time: 3:48:09
```

**Review by twhetzel — CHANGES_REQUESTED** [review]
The full build no longer works after further changes were added after this was already reviewed and previously approved.

The GH Action in the Mondo repo "Update Externally Provided Content" does still work so this PR will be left unmerged until the issue with the full build can be sorted out and fixed.

**Review by joeflack4 — CHANGES_REQUESTED** [review]
Build is failing.

File: src/ontology/config/robot_report_external_content.txt
Comment (twhetzel) [inline src/ontology/config/robot_report_external_content.txt:6]: Consider removing this check since it looks like it is covered by the default ROBOT check of `duplicate_scoped_synonym` already included in this file.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [inline src/ontology/mondo-ingest.Makefile:626]: ## Bug: `DOWNLOAD_EXTERNAL=false`
[Trish commented](https://github.com/monarch-initiative/mondo-ingest/pull/628#issuecomment-2334371732) that  `sh run.sh make update-externally-managed-content -B`, which currently includes the recent simulation goal that Nico added, is erroring: 

```
tmp/external/processed-efo-proxy-merges.robot.owl (No such file or directory)
```

The problem involves running some `make` goals in a clone of `mondo` within `mondo-ingest`'s `tmp/` dir. There is an option `DOWNLOAD_EXTERNAL` (`true` by default in the `mondo.Makefile`), which downloads files from `mondo-ingest`. However, during this simulation goal, it passes `DOWNLOAD_EXTERNAL=false`. I think that Nico added that to avoid a kind of circular dependency, as it actually downloads these files from the `mondo-ingest` repo. However, if it does not download these files, then they will not exist, causing this error.

Fixed by:
- #672 

Though I don't know if there's a good reason why Nico passed `DOWNLOAD_EXTERNAL=false`. Maybe there's a different solution he would prefer.

**Comment by twhetzel** [comment]
@matentzn can you confirm this feature branch still builds and merge into `develop`? Last time I tried to run the build from this branch it did not work.

**Comment by matentzn** [comment]
See slack need to merge develop into main first

**Comment by twhetzel** [comment]
@matentzn will the [EMC GitHub Action](https://github.com/monarch-initiative/mondo/blob/master/.github/workflows/external_content.yaml) need to be updated once this is merged? It does not look like it will need any updates, but want to confirm.

**Comment by twhetzel** [comment]
This "banana" `MONDO:MONDO:0859214` is causing the issue. Another QC check will be added to address this.

**Comment by twhetzel** [comment]
@matentzn is this ready for re-review?

File: src/ontology/external/nord-qc-failures.md
Comment (matentzn) [inline src/ontology/external/nord-qc-failures.md:14]: FYI @twhetzel this is where the Banana was identified and will be reported to our stakeholders.

Comment (matentzn) [reply src/ontology/external/nord-qc-failures.md:14]: This is a consequence of fixing the issue that causes the build file

https://github.com/monarch-initiative/mondo-ingest/pull/628#issuecomment-2481050169

**Comment by matentzn** [comment]
@twhetzel this is ready for review and test build.

File: src/ontology/config/robot_report_external_content.txt
Comment (twhetzel) [reply src/ontology/config/robot_report_external_content.txt:6]: Agree!

**Review by twhetzel — APPROVED** [review]
LGTM

**Pull request merged**
